### PR TITLE
Enables private git repository python packages

### PR DESCRIPTION
**Pull request by johnbaldwin**
from a deployment configuration perspective, behaves the same as
'EDXAPP_EXTRA_REQUIREMENTS'. Specify your private repositories in
'EDXAPP_PRIVATE_REQUIREMENTS'

This means we can do away with the `private_requirements.txt` file if

Preparation for upstream: I added a check in this task so that if the `EDXAPP_PRIVATE_REQUIREMENTS` is not specified in the vars, but private requirements are specified, then it doesn't try to create the file. This is so that the deploy doesn't break under this condition. Also the task will not create the private requirements file if `EDXAPP_INSTALL_PRIVATE_REQUIREMENTS` is not defined or not set

### CODE REVIEW

**Comment by OmarIthawi** [comment]
Thanks @johnbaldwin, this looks like a fairly contributable feature. Did you consider pushing this upstream?

If not please consider adding it to this [dedicated Trello card](https://trello.com/c/PBDK9NZG) so at least someone else will.

**Comment by tkeemon** [comment]
@johnbaldwin thanks for this. This solves a huge pain point in our deployments.

One question: I'm assuming that each private repo needs to be connected to homerbot for this to work. Where is homerbot's private key being held?

**Comment by melvinsoft** [comment]
@johnbaldwin I'm wondering the same as TJ is asking.

**Comment by johnbaldwin** [comment]
[(1) Figure out how to improve private_requirements](https://trello.com/c/GJquSDrj/178-1-figure-out-how-to-improve-privaterequirements)

File: playbooks/roles/edxapp/templates/private_requirements.txt.j2 [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# This file was automatically generated by Open edX Configuration.
```
Comment (johnbaldwin) [inline playbooks/roles/edxapp/templates/private_requirements.txt.j2]: @OmarIthawi @abeals @bryanlandia I figure you guys would have the strongest opinion here. Do you think this is ok or do you think we should change or add to the message?

**Comment by johnbaldwin** [comment]
All this PR does is automatically generate the private_requirements.txt file from settings in the server-vars.yml (or really any added vars file)

the SSH key structure is already in place from edX.org from what I've read over the code in edxapp/tasks/deploy.yml

As far as the git SSH key requirements: All that is required is that the account tied to the SSH key pair has read access to multiple repositories. This is why a deploy key won't work.

So far, we've followed a convention of unique keypairs per customer. I think we are also following unique keypairs per deployment.


I think at a future date it might be good to do a flexible system that *can* bind a keypair identity to each private git repo. So that 'homerbot' might be used for repos A, B, C and a deploy key for repo D or if each repo can have it's own git identify, then we can go to using deploy keys, which I think is more ideal, but more work

**Comment by johnbaldwin** [comment]
@OmarIthawi So to answer your question:

>  the plan is to make a new homer private key for each client, and add it home's GitHub. Right 

Yes, that's our plan. This should be the case for each of the configs that already has private_requirements.txt files in the deployment's files dir. We can check with the MD5 signatures of the public keys

**Comment by johnbaldwin** [comment]
@tkeemon 

> One question: I'm assuming that each private repo needs to be connected to homerbot for this to work. Where is homerbot's private key being held?

Yes, the private keys are in the secret.yml files.  You give me (or Morgan or Aaron) the public key and one of us adds to Homerbot's account. I want to automate this if github has an API to manage SSH keys. If GH does not have an API to automate key management, then I want to do a mini-project to make the private repos each have it's own deploy key and either have global deploy keys, or the set of unique deploy keys per deployment. Again, need to be able to automate this via GH API

**Comment by OmarIthawi** [comment]
There's a standard'ish Ansible managed message. You could use that.

**Comment by johnbaldwin** [comment]
@OmarIthawi I added changes to make this more robust. I also looked at the existing tests and doesn't look like there are any significant tests, just three specialized test shell scripts

**Comment by tkeemon** [comment]
@OmarIthawi but doesn't homerbot's private key need to exist somewhere in server-vars? 

I like this approach, but I guess my overall questions are: 

- What is the workflow for using this approach? 
- How do I add a private key to Homerbot? 
- How do I connect Homerbot with a private repo? 

Some docs for this procedure would be really helpful.

File: playbooks/roles/edxapp/templates/private_requirements.txt.j2 (hunk 1 above)
Comment (tkeemon) [reply playbooks/roles/edxapp/templates/private_requirements.txt.j2]: @johnbaldwin I think this is the standard way to generate that message: 

https://github.com/appsembler/configuration/blob/9e9a467f1cf0bfd8b06a926747d40ccd8f2a9d67/playbooks/roles/edx_service/templates/config.yml.j2#L2

**Comment by OmarIthawi** [comment]
@johnbaldwin three points: 

 - [ ] I second @tkeemon's idea, Homerbot creds should be in LastPass. 

 - [ ] Shouldn't we have another bot for private repos? We could call it `@margebot`.

 - [ ] However, homerbot access rights should be very limited know. Since it (the bot) will have its private keys all over the places, it would be a weak point.
    At the very least it (the bot) shouldn't have access to `edx-configs` and `edx-terraform`. But whitelisting is better than blacklisting.

**Comment by tkeemon** [comment]
@OmarIthawi I agree with your points #1 and 3. But Homerbot already deals with private repos, so we'll need to find another task for Margebot.

**Comment by johnbaldwin** [comment]
@OmarIthawi @tkeemon @melvinsoft @bryanlandia @amirtds 

First, here's a doc to shed a lot more light on Homerbot: https://docs.google.com/document/d/1KmpVBzYSHzXT4toFfFiN29T2mIs7iuRHEwtj9ad_zmM

/cc @abeals 
* The purpose of homerbot is for private repos because of the limitation of deploy keys (please read the doc I linked above)

* Homerbot has read only access to repositories for the repositories for which I have admin access and can confirm. I can't confirm edx-themes-codebase

* I added Homerbot password to keybase to restrict access to Black Team group. Caveat here is that we use 2FA for auth so there's some hoops for you guys to log in as homerbot in github (See the doc I linked above for details). But upon consideration, I don't think we'll need to log in or create lots of SSH keys (Please read my comment in the security section of the doc)

**Comment by OmarIthawi** [comment]
@johnbaldwin Please address TJ's comment and the PR should be good to go! Let's not keep it hanging for a long while. This is a very useful PR IMHO.

I'd say go ahead and merged it! If it doesn't go well, we can always revert!

**Comment by OmarIthawi** [comment]
@johnbaldwin let's get this merged.

**Pull request merged**